query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+            inner_attributes->emplace(attr);
+          }
+
+          non_hash_join_predicates->emplace_back(
+              std::static_pointer_cast<const E::Predicate>(new_expression));
+          return E::ExpressionPtr();
+        }
+      }
+      return expression;
+    }
+    default: {
+      validateNonOuterAttributeReference(expression);
+      return expression;
+    }
+  }
+}
+
+void UnnestSubqueriesForNonRootLogical::validateNonOuterAttributeReference(
+    const E::ExpressionPtr &expression) {
+  const std::vector<E::AttributeReferencePtr> referenced_attributes =
+      expression->getReferencedAttributes();
+  for (const E::AttributeReferencePtr &referenced_attribute : referenced_attributes) {
+    if (isCorrelatedOuterAttribute(referenced_attribute)) {
+      THROW_SQL_ERROR()
+          << "Outer attribute reference can only appear in a single-attribute equality predicate";
+    }
+  }
+}
+
+bool UnnestSubqueriesForNonRootLogical::isCorrelatedOuterAttribute(
+    const E::AttributeReferencePtr &attribute) const {
+  return attribute->scope() == E::AttributeReferenceScope::kOuter
+         && uncorrelated_subqueries_->find(attribute->id()) == uncorrelated_subqueries_->end();
+}
+
+L::LogicalPtr UnnestSubqueriesForNonRootLogical::eliminateNestedScalarQueries(const L::LogicalPtr &node) {
+  const std::vector<E::ExpressionPtr> &input_expressions = node->input_expressions();
+  const std::vector<E::AttributeReferencePtr> input_attributes =
+      node->getInputAttributes();
+  E::UnorderedNamedExpressionSet input_attribute_set(input_attributes.begin(),
+                                                     input_attributes.end());
+
+  // Transform subquery experssions into non-subquery expressions.
+  std::vector<CorrelatedQueryInfo> correlated_query_info_vec;
+  std::vector<E::ExpressionPtr> new_input_expressions;
+  bool has_changed_expression = false;
+  for (const E::ExpressionPtr &input_expression : input_expressions) {
+//    std::cerr << "to unnest expr: " << input_expression->toString();
+    std::vector<E::PredicatePtr> correlated_predicates;
+    UnnestSubqueriesForExpession unnest_rule_for_expr(
+        input_attribute_set,
+        context_,
+        uncorrelated_subqueries_,
+        &correlated_query_info_vec);
+    const E::ExpressionPtr new_input_expression =
+        unnest_rule_for_expr.apply(input_expression);
+    if (new_input_expression != input_expression && !has_changed_expression) {
+      has_changed_expression = true;
+    }
+    new_input_expressions.emplace_back(new_input_expression);
+  }
+
+  if (has_changed_expression) {
+    // If there are correlated subquery expressions, add logical joins.
+    if (!correlated_query_info_vec.empty()) {
+      L::LogicalPtr new_child;
+
+      if (new_input_expressions[0] == nullptr
+          && node->getLogicalType() == L::LogicalType::kNestedLoopsJoin) {
+        new_child = node;
+      } else {
+        DCHECK_EQ(node->children().size(), 1u);
```

Comment:
Do you remember that we put the expected value in the left hand side, while the checked on in the right?

`DCHECK_EQ(1u, node->children().size());`